## dataset statistic ##

In [1]:
"""
    dataset information:
        train: 29837
        dev: 4475
        test: 2985
"""

import pandas as pd
import json


# training data

with open('dataset/MathQA/train.json') as f:
    data = json.loads(f.read())
    
train_df = pd.DataFrame(data)

print('training problems number: ', train_df['Problem'].count())
print(train_df.groupby(['category'])['Problem'].count())

train_df.head(5)

D:\Program Files (x86)\Anaconda3\envs\pytorch\lib\site-packages\numpy\_distributor_init.py:32: UserWarning: loaded more than 1 DLL from .libs:
D:\Program Files (x86)\Anaconda3\envs\pytorch\lib\site-packages\numpy\.libs\libopenblas.PYQHXLVVQ7VESDPUVUADXEVJOBGHJPAY.gfortran-win_amd64.dll
D:\Program Files (x86)\Anaconda3\envs\pytorch\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
  stacklevel=1)


training problems number:  29837
category
gain            5120
general        13273
geometry        2117
other           1814
physics         7063
probability      450
Name: Problem, dtype: int64


Problem  \
0  the banker ' s gain of a certain sum due 3 yea...   
1  average age of students of an adult school is ...   
2  sophia finished 2 / 3 of a book . she calculat...   
3                        120 is what percent of 50 ?   
4  there are 10 girls and 20 boys in a classroom ...   

                                           Rationale  \
0  "explanation : t = 3 years r = 10 % td = ( bg ...   
1  "explanation : let the original no . of studen...   
2  let xx be the total number of pages in the boo...   
3  "50 * x = 120 - - > x = 2.4 - - > 2.4 expresse...   
4  if girls is 10 and boys is 20 , then 10 / 20 ....   

                                             options correct  \
0  a ) rs . 400 , b ) rs . 300 , c ) rs . 500 , d...       a   
1  a ) 1200 , b ) 120 , c ) 360 , d ) 240 , e ) n...       d   
2    a ) 229 , b ) 270 , c ) 877 , d ) 266 , e ) 281       b   
3  a ) 5 % , b ) 240 % , c ) 50 % , d ) 2 % , e )...       b   
4  a ) 1 / 2 , b ) 1 / 3 , c ) 1 / 5 , d ) 10 / 3...       a   

                                   annotated_formula  \
0  divide(multiply(const_100, divide(multiply(36,...   
1  multiply(divide(subtract(multiply(add(32, 4), ...   
2        divide(90, subtract(const_1, divide(2, 3)))   
3               multiply(divide(120, 50), const_100)   
4                                     divide(10, 20)   

                                      linear_formula category  
0  multiply(n2,const_100)|multiply(n0,n1)|divide(...     gain  
1  add(n2,n3)|multiply(n1,n2)|multiply(n1,#0)|sub...  general  
2   divide(n0,n1)|subtract(const_1,#0)|divide(n2,#1)  general  
3              divide(n0,n1)|multiply(#0,const_100)|     gain  
4                                      divide(n0,n1)    other

In [2]:
# dev data

with open('dataset/MathQA/dev.json') as f:
    data = json.loads(f.read())
    
dev_df = pd.DataFrame(data)

print('dev problems number: ', dev_df['Problem'].count())
print(dev_df.groupby(['category'])['Problem'].count())

dev_df.head(5)

dev problems number:  4475
category
gain            743
general        2006
geometry        308
other           288
physics        1056
probability      74
Name: Problem, dtype: int64


Problem  \
0  a multiple choice test consists of 4 questions...   
1  a 3 - digit positive integer is chosen at rand...   
2  if x and y are positive integers and 7 + x + y...   
3  the hcf and lcm of two numbers m and n are res...   
4  in a kilometer race , a beats b by 48 meters o...   

                                           Rationale  \
0  "5 choices for each of the 4 questions , thus ...   
1  p ( three odd digits ) = 5 / 9 * 5 / 10 * 5 / ...   
2  "try each answer choices . for a : 7 + 3 + xy ...   
3  answer we have , m x n = 6 x 210 = 1260 ∴ 1 / ...   
4  "time taken by b run 1000 meters = ( 1000 * 12...   

                                             options correct  \
0    a ) 24 , b ) 120 , c ) 625 , d ) 720 , e ) 1024       c   
1  a ) 1 / 2 , b ) 31 / 36 , c ) 49 / 54 , d ) 7 ...       b   
2              a ) 3 , b ) 5 , c ) 6 , d ) 8 , e ) 9       c   
3  a ) 1 / 35 , b ) 3 / 35 , c ) 5 / 37 , d ) 2 /...       d   
4  a ) 238 sec , b ) 190 sec , c ) 667 sec , d ) ...       a   

                                   annotated_formula  \
0                                        power(5, 4)   
1  subtract(const_1, power(divide(const_1, const_...   
2                                   add(21, const_1)   
3                       divide(72, multiply(6, 210))   
4  subtract(divide(multiply(const_1, const_1000),...   

                                      linear_formula category  
0                                      power(n1,n0)|  general  
1  divide(const_1,const_2)|power(#0,n0)|subtract(...  general  
2                                   add(n1,const_1)|  general  
3                      multiply(n0,n1)|divide(n2,#0)  general  
4  divide(n0,n1)|multiply(const_1,const_1000)|div...  physics

In [3]:
# test data

with open('dataset/MathQA/test.json') as f:
    data = json.loads(f.read())
    
test_df = pd.DataFrame(data)

print('test problems number: ', test_df['Problem'].count())
print(test_df.groupby(['category'])['Problem'].count())

test_df.head(5)

test problems number:  2985
category
gain            558
general        1301
geometry        200
other           172
physics         702
probability      52
Name: Problem, dtype: int64


Problem  \
0  a shopkeeper sold an article offering a discou...   
1  what will be the difference between simple and...   
2  there are 28 stations between hyderabad and ba...   
3  the present population of a town is 3888 . pop...   
4                 the triplicate ratio of 1 : 9 is ?   

                                           Rationale  \
0  "giving no discount to customer implies sellin...   
1  "s . i . = ( 1000 * 14 * 4 ) / 100 = rs . 560 ...   
2  "the total number of stations = 30 from 30 sta...   
3  "p = 3888 r = 20 % required population of town...   
4                     "13 : 93 = 1 : 729 answer : e"   

                                             options correct  \
0  a ) 38 , b ) 27.675 , c ) 30 , d ) data inadeq...       a   
1    a ) 129 , b ) 130 , c ) 124 , d ) 133 , e ) 145       a   
2    a ) 156 , b ) 167 , c ) 870 , d ) 352 , e ) 380       c   
3  a ) 2500 , b ) 2100 , c ) 3500 , d ) 3600 , e ...       e   
4  a ) 1 : 0 , b ) 1 : 8 , c ) 1 : 7 , d ) 1 : 2 ...       e   

                                   annotated_formula  \
0  subtract(divide(multiply(add(const_100, 31.1),...   
1  subtract(subtract(multiply(1000, power(add(div...   
2  multiply(add(28, const_1), add(add(28, const_1...   
3   add(3888, divide(multiply(3888, 20), const_100))   
4   divide(power(const_2.0, 9), power(const_3.0, 9))   

                                      linear_formula category  
0  add(n1,const_100)|subtract(const_100,n0)|multi...     gain  
1  divide(n0,const_100)|add(#0,const_1)|multiply(...     gain  
2   add(n0,const_1)|add(#0,const_1)|multiply(#0,#1)|  physics  
3   multiply(n0,n1)|divide(#0,const_100)|add(n0,#1)|     gain  
4  power(const_2.0,n1)|power(const_3.0,n1)|divide...    other

### training data information:

##### count: 29837

    gain          5120
    general       13273
    geometry      2117
    other         1814
    physics       7063
    probability   450

In [85]:
import spacy

spacy_en = spacy.load('en_core_web_sm')

def tokenize_en(text):
    return [tok.text for tok in spacy_en.tokenizer(text)]

def split_formula(text):
    return text.split('|')

train_src = train_df['Problem'].apply(tokenize_en)
print(train_src)

train_trg = train_df['linear_formula'].apply(split_formula)
print(train_trg)

0        [the, banker, ', s, gain, of, a, certain, sum,...
1        [average, age, of, students, of, an, adult, sc...
2        [sophia, finished, 2, /, 3, of, a, book, ., sh...
3                      [120, is, what, percent, of, 50, ?]
4        [there, are, 10, girls, and, 20, boys, in, a, ...
                               ...                        
29832    [a, man, invests, some, money, partly, in, 10,...
29833    [average, of, 10, matches, is, 32, ,, how, man...
29834       [solve, below, question, 2, x, +, 1, =, -, 23]
29835                     [calculate, 85184, ÷, ?, =, 352]
29836    [7.51, 8.22, 7.86, 8.36, 8.09, 7.83, 8.30, 8.0...
Name: Problem, Length: 29837, dtype: object
0        [multiply(n2,const_100), multiply(n0,n1), divi...
1        [add(n2,n3), multiply(n1,n2), multiply(n1,#0),...
2        [divide(n0,n1), subtract(const_1,#0), divide(n...
3                [divide(n0,n1), multiply(#0,const_100), ]
4                                          [divide(n0,n1)]
            

## calculate the max length of the problem sentences ##

In [86]:
max_len = 0

for i in train_src:
    if max_len < len(i):
        max_len = len(i) 
print("max length in tokens: ", max_len)


for sent in train_df['Problem']:
    
    # Tokenize the text and add `[CLS]` and `[SEP]` tokens.
    input_ids = tokenizer.encode(sent, add_special_tokens=True)

    # Update the maximum sentence length.
    max_len = max(max_len, len(input_ids))
    
print('Max sentence length: ', max_len)

max length in tokens:  356
Max sentence length:  373


In [87]:
max_len = 0
for i in train_trg:
    if max_len < len(i):
        max_len = len(i)
        
print("max length in target tokens: ", max_len)

max length in target tokens:  55


In [7]:
from torchtext.legacy import data

SRC = data.Field(tokenize = tokenize_en, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True)

TRG = data.Field(tokenize = split_formula, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True)

In [ ]:
# data_path = 'dataset/MathQA/'
# train_data, valid_data, test_data = data.TabularDataset.splits(
#     path=data_path,
#     train='train.json',
#     validation='dev.json',
#     test='test.json',
#     format='json',
#     fields=[('SRC', SRC), ('TRG', TRG)],
# )

## tokenization ##

In [58]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

encoded_corpus = tokenizer(
    text=train_df['Problem'].tolist(),
    padding='max_length',
    truncation='longest_first',
    max_length=400,
    return_attention_mask=True
)

In [ ]:
encoded_corpus_trg = tokenizer(
    text=train_df['Problem'].tolist(),
    padding='max_length',
    truncation='longest_first',
    max_length=400,
    return_attention_mask=True
)

In [59]:
input_ids = encoded_corpus['input_ids']
attention_mask = encoded_corpus['attention_mask']

In [60]:
print(train_df['Problem'][0])
print(len(input_ids[0]))

the banker ' s gain of a certain sum due 3 years hence at 10 % per annum is rs . 36 . what is the present worth ?
400


In [54]:
# inputs = tokenizer(train_df['Problem'][0], padding="max_length", truncation=True)
# inputs

## input formatting ##

In [65]:
from sklearn.model_selection import train_test_split

test_size = 0.1
seed = 2021

train_inputs, test_inputs = train_test_split(
    input_ids, 
#     labels, 
    test_size=test_size, 
    random_state=seed
)

train_masks, test_masks = train_test_split(
    attention_mask, 
#     labels, 
    test_size=test_size, 
    random_state=seed
)

In [66]:
print(train_inputs[0])

[101, 24181, 11687, 4847, 2373, 128, 120, 1492, 24438, 1104, 170, 1520, 1107, 122, 188, 1204, 1989, 1105, 126, 120, 130, 24438, 1104, 1103, 2735, 1520, 1107, 123, 183, 1181, 1989, 119, 1191, 1175, 1127, 5306, 5097, 8362, 11613, 1170, 123, 183, 1181, 1989, 117, 1293, 1242, 5097, 1127, 1175, 1107, 1103, 1520, 136, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [67]:
import torch
from torch.utils.data import TensorDataset, DataLoader

batch_size = 32

def create_dataloaders(inputs, masks, batch_size):
    input_tensor = torch.tensor(inputs)
    mask_tensor = torch.tensor(masks)
#     labels_tensor = torch.tensor(labels)
    dataset = TensorDataset(input_tensor, mask_tensor)
    dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
    return dataloader

train_dataloader = create_dataloaders(train_inputs, train_masks, batch_size)

test_dataloader = create_dataloaders(test_inputs, test_masks, batch_size)

In [75]:
train_dataloader

In [69]:
from transformers import AutoModelForSequenceClassification

model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased", num_labels=2)

Some weights of the model checkpoint at bert-base-cased were not used when initializing BertForSequenceClassification: ['cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at b

In [70]:
from transformers import AdamW

optimizer = AdamW(model.parameters(), lr=5e-5)

In [71]:
from transformers import get_scheduler

num_epochs = 3
num_training_steps = num_epochs * len(train_dataloader)
lr_scheduler = get_scheduler(
    "linear",
    optimizer=optimizer,
    num_warmup_steps=0,
    num_training_steps=num_training_steps
)

In [72]:
import torch

device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
model.to(device)

BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(28996, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0): BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, element

In [88]:
# from tqdm.auto import tqdm

# progress_bar = tqdm(range(num_training_steps))

# model.train()

# for epoch in range(num_epochs):
#     for batch in train_dataloader:
#         batch = {k: v.to(device) for k, v in batch.items()}
#         outputs = model(**batch)
#         loss = outputs.loss
#         loss.backward()

#         optimizer.step()
#         lr_scheduler.step()
#         optimizer.zero_grad()
#         progress_bar.update(1)

In [ ]:
metric= load_metric("accuracy")
model.eval()
for batch in eval_dataloader:
    batch = {k: v.to(device) for k, v in batch.items()}
    with torch.no_grad():
        outputs = model(**batch)

    logits = outputs.logits
    predictions = torch.argmax(logits, dim=-1)
    metric.add_batch(predictions=predictions, references=batch["labels"])

metric.compute()

## Implementing the model in PyTorch ##

In [76]:
import torch.nn as nn
from transformers import CamembertModel, BertModel


class CamembertRegressor(nn.Module):
    
    def __init__(self, drop_rate=0.2, freeze_camembert=False):
        super(CamembertRegressor, self).__init__()
        D_in, D_out = 768, 1
        
        self.camembert = CamembertModel.from_pretrained('camembert-base')
        self.regressor = nn.Sequential(nn.Dropout(drop_rate),nn.Linear(D_in, D_out))
        
        
    def forward(self, input_ids, attention_masks):
        outputs = self.camembert(input_ids, attention_masks)
        class_label_output = outputs[1]
        outputs = self.regressor(class_label_output)    
        return outputs
    
    
model = CamembertRegressor(drop_rate=0.2)

Downloading:   0%|          | 0.00/508 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/424M [00:00<?, ?B/s]

Some weights of the model checkpoint at camembert-base were not used when initializing CamembertModel: ['lm_head.decoder.weight', 'lm_head.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.layer_norm.bias']
- This IS expected if you are initializing CamembertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing CamembertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [77]:
import torch

if torch.cuda.is_available():       
    device = torch.device("cuda")
    print("Using GPU.")
else:
    print("No GPU available, using the CPU instead.")
    device = torch.device("cpu")

model.to(device)

No GPU available, using the CPU instead.


CamembertRegressor(
  (camembert): CamembertModel(
    (embeddings): RobertaEmbeddings(
      (word_embeddings): Embedding(32005, 768, padding_idx=1)
      (position_embeddings): Embedding(514, 768, padding_idx=1)
      (token_type_embeddings): Embedding(1, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): RobertaEncoder(
      (layer): ModuleList(
        (0): RobertaLayer(
          (attention): RobertaAttention(
            (self): RobertaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): RobertaSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): Laye

In [78]:
from transformers import AdamW

optimizer = AdamW(model.parameters(), lr=5e-5, eps=1e-8)

In [79]:
from transformers import get_linear_schedule_with_warmup

epochs = 5
total_steps = len(train_dataloader) * epochs
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=total_steps)

In [80]:
loss_function = nn.MSELoss()

In [82]:
from torch.nn.utils.clip_grad import clip_grad_norm

def train(model, optimizer, scheduler, loss_function, epochs, train_dataloader, device, clip_value=2):
    
    for epoch in range(epochs):
        print(epoch)
        print("-----")
        best_loss = 1e10
        model.train()
        for step, batch in enumerate(train_dataloader): 
            print(step)  
            batch_inputs, batch_masks = tuple(b.to(device) for b in batch)
            model.zero_grad()
            outputs = model(batch_inputs, batch_masks)           
            loss = loss_function(outputs.squeeze())
            loss.backward()
            clip_grad_norm(model.parameters(), clip_value)
            optimizer.step()
            scheduler.step()
            
    return model


model = train(model, optimizer, scheduler, loss_function, epochs, train_dataloader, device, clip_value=2)

0
-----
0


TypeError: forward() missing 1 required positional argument: 'target'